### PR TITLE
Added MasterVlanPlugin struct to nad package

### PR DESCRIPTION
**Pull request by kononovn**

### CODE REVIEW

File: pkg/nad/masterplugin.go [hunk 1]
```diff
@@ -174,10 +175,106 @@ func (plugin *MasterBridgePlugin) WithIPAM(ipam *IPAM) *MasterBridgePlugin {
...
+			Name:       name,
+			Type:       "vlan",
+			VlanID:     vlanID,
+		},
+	}
+
+	if vlanID > 4094 {
+		glog.V(100).Infof("error vlan id can not be greater than 4094")
+
+		builder.errorMsg = "MasterVlanPlugin vlanId is greater than 4094"
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:209]: vlanID

File: pkg/nad/masterplugin.go [hunk 2]
```diff
@@ -174,10 +175,106 @@ func (plugin *MasterBridgePlugin) WithIPAM(ipam *IPAM) *MasterBridgePlugin {
...
+	}
+
+	if builder.masterPlugin.Name == "" {
+		glog.V(100).Infof("error MasterVlanPlugin can not be empty")
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:213]: MasterVlanPlugin name

File: pkg/nad/masterplugin.go [hunk 3]
```diff
@@ -174,10 +175,106 @@ func (plugin *MasterBridgePlugin) WithIPAM(ipam *IPAM) *MasterBridgePlugin {
...
+
+		builder.errorMsg = "MasterVlanPlugin name is empty"
+	}
+
+	return &builder
+}
+
+// WithIPAM defines IPAM configuration to MasterVlanPlugin. Default is empty.
+func (plugin *MasterVlanPlugin) WithIPAM(ipam *IPAM) *MasterVlanPlugin {
+	glog.V(100).Infof("Adding ipam configuration %v to MasterVlanPlugin", ipam)
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:223]: 1) IPAM configuration
2) ipam is a struct, so I think better is:
` glog.V(100).Infof("Adding IPAM configuration  to MasterVlanPlugin: %v", ipam)`

File: pkg/nad/masterplugin.go [hunk 4]
```diff
@@ -174,10 +175,106 @@ func (plugin *MasterBridgePlugin) WithIPAM(ipam *IPAM) *MasterBridgePlugin {
...
+
+	if plugin.masterPlugin == nil {
+		glog.V(100).Infof(msg.UndefinedCrdObjectErrString("MasterVlanPlugin"))
+		plugin.errorMsg = msg.UndefinedCrdObjectErrString("MasterVlanPlugin")
+	}
+
+	if ipam == nil {
+		glog.V(100).Infof("error adding empty ipam to MasterVlanPlugin")
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:231]: "error adding empty ipam to MasterVlanPlugin"

File: pkg/nad/masterplugin.go [hunk 5]
```diff
@@ -174,10 +175,106 @@ func (plugin *MasterBridgePlugin) WithIPAM(ipam *IPAM) *MasterBridgePlugin {
...
+		plugin.errorMsg = invalidIpamParameterMsg
+	}
+
+	plugin.masterPlugin.Ipam = ipam
+
+	return plugin
+}
+
+// WithMasterInterface defines master interface to MasterVlanPlugin. Default is cn0.
+func (plugin *MasterVlanPlugin) WithMasterInterface(master string) *MasterVlanPlugin {
```
Comment (evgenLevin) [inline pkg/nad/masterplugin.go:242]: (masterInterfaceName string)

**Review by evgenLevin — COMMENTED** [review]
A few comments

**Comment by kononovn** [comment]
@evgenLevin PTAL

File: pkg/nad/masterplugin.go [hunk 6]
```diff
@@ -174,10 +175,106 @@ func (plugin *MasterBridgePlugin) WithIPAM(ipam *IPAM) *MasterBridgePlugin {
...
+	if plugin.masterPlugin == nil {
+		glog.V(100).Infof(msg.UndefinedCrdObjectErrString("MasterVlanPlugin"))
+		plugin.errorMsg = msg.UndefinedCrdObjectErrString("MasterVlanPlugin")
+	}
+
+	if masterInterfaceName == "" {
+		glog.V(100).Infof("error to add masterInterfaceName interface, the name of interface can not be empty")
+
+		plugin.errorMsg = "invalid masterInterfaceName parameter"
+	}
```
Comment (trewest) [inline pkg/nad/masterplugin.go:254]: shouldn't you check if errorMsg != "" and return here?

File: pkg/nad/masterplugin.go [hunk 7]
```diff
@@ -174,10 +175,106 @@ func (plugin *MasterBridgePlugin) WithIPAM(ipam *IPAM) *MasterBridgePlugin {
...
+		glog.V(100).Infof(msg.UndefinedCrdObjectErrString("MasterVlanPlugin"))
+		plugin.errorMsg = msg.UndefinedCrdObjectErrString("MasterVlanPlugin")
+	}
+
+	if ipam == nil {
+		glog.V(100).Infof("error adding empty ipam to MasterVlanPlugin")
+
+		plugin.errorMsg = invalidIpamParameterMsg
+	}
+
```
Comment (trewest) [inline pkg/nad/masterplugin.go:235]: shouldn't you check if errorMsg != "" and return here?

File: pkg/nad/masterplugin.go (hunk 6 above)
Comment (kononovn) [reply pkg/nad/masterplugin.go:254]: No I have this check before I return ready struct in GetMasterPluginConfig

File: pkg/nad/masterplugin.go (hunk 7 above)
Comment (kononovn) [reply pkg/nad/masterplugin.go:235]: No I have this check before I return ready struct in GetMasterPluginConfig

File: pkg/nad/masterplugin.go [hunk 8]
```diff
@@ -174,10 +175,106 @@ func (plugin *MasterBridgePlugin) WithIPAM(ipam *IPAM) *MasterBridgePlugin {
...
+	return plugin
+}
+
+// WithLinkInContainer defines MasterVlanPlugin using linkInContainer feature.
+func (plugin *MasterVlanPlugin) WithLinkInContainer() *MasterVlanPlugin {
+	if plugin.masterPlugin == nil {
+		glog.V(100).Infof(msg.UndefinedCrdObjectErrString("MasterVlanPlugin"))
+		plugin.errorMsg = msg.UndefinedCrdObjectErrString("MasterVlanPlugin")
+	}
+
```
[the remark this answers is not in the document]
Comment (kononovn) [reply pkg/nad/masterplugin.go:267]: No I have this check before I return ready struct in GetMasterPluginConfig

**Review by trewest — CHANGES_REQUESTED** [review]
Can you address the following comments. It seems standard in other pkgs that we refrain from updating the builder if errorMsg != ""

File: pkg/nad/masterplugin.go (hunk 7 above)
Comment (kononovn) [reply pkg/nad/masterplugin.go:235]: Added condition

**Comment by kononovn** [comment]
@trewest Ok I added required validation to my functs. PTAL

**Review by trewest — APPROVED** [review]
/lgtm

**Pull request merged**